keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
+    return deliveryConfig.environments.map { environment ->
+      val artifactVersions = deliveryConfig.artifacts.map { artifact ->
+        val versionsInEnvironment = jooq
+          .select(
+            ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION,
+            ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT,
+            ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT
+          )
+          .from(
+            ENVIRONMENT_ARTIFACT_VERSIONS,
+            DELIVERY_ARTIFACT,
+            ENVIRONMENT,
+            DELIVERY_CONFIG
+          )
+          .where(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID))
+          .and(DELIVERY_ARTIFACT.NAME.eq(artifact.name))
+          .and(DELIVERY_ARTIFACT.TYPE.eq(artifact.type.name))
+          .and(ENVIRONMENT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID))
+          .and(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(DELIVERY_CONFIG.UID))
+          .and(DELIVERY_CONFIG.NAME.eq(deliveryConfig.name))
+          .and(ENVIRONMENT.NAME.eq(environment.name))
+        val pendingVersions = jooq
```

Comment:
This needs to take artifact status into account since environments can be concerned with different sets of statuses within a delivery config. The current implementation would list `CANDIDATE` artifacts as pending on an environment that only accepts `RELEASE`.